src/amber/router/cookies.cr

```diff
@@ -0,0 +1,217 @@
+      include Enumerable(String)
+      include ChainedCookieStore
+
+      def self.from_headers(headers)
+        cookies = {} of String => HTTP::Cookie
+        if values = headers.get?("Cookie")
+          values.each do |header|
+            HTTP::Cookie::Parser.parse_cookies(header) do |cookie|
+              cookies[cookie.name] = cookie
+            end
+          end
+        end
+        cookies
+      end
+
+      def self.build(request, key_generator = Amber::Support::KeyGenerator.new("secret"))
+        headers = request.headers
+        host = request.host
+        secure = request.ssl?
+
+        new(key_generator, host, secure).tap do |jar|
+          jar.update(from_headers(headers))
+        end
+      end
+
+      getter :cookies
+      property :host
+      property :secure
+
+      def initialize(@key_generator, @host = nil, @secure = nil)
+        @cookies = {} of String => String
+        @set_cookies = {} of String => HTTP::Cookie
+        @delete_cookies = {} of String => HTTP::Cookie
+      end
+
+      def update(cookies)
+        cookies.each do |name, cookie|
+          @cookies[name] = cookie.value
+        end
+      end
+
+      def each(&block : T -> _)
+        @cookies.values.each do |cookie|
+          yield cookie
+        end
+      end
+
+      def each
+        @cookies.each_value
+      end
+
+      def [](name)
+        get(name)
+      end
+
+      def get(name)
+        @cookies[name]?
+      end
+
+      def set(name : String, value : String, path : String = "/",
+              expires : Time? = nil, domain : String? = nil,
+              secure : Bool = false, http_only : Bool = false,
+              extension : String? = nil)
+        if @cookies[name]? != value || expires
+          @cookies[name] = value
+          @set_cookies[name] = HTTP::Cookie.new(name, value, path, expires, domain, secure, http_only, extension)
+          @delete_cookies.delete(name) if @delete_cookies.has_key?(name)
+        end
+      end
+
+      def delete(name : String, path = "/", domain : String? = nil)
+        return unless @cookies.has_key?(name)
+
+        value = @cookies.delete(name)
+        @delete_cookies[name] = HTTP::Cookie.new(name, "", path, Time.epoch(0), domain)
+        value
+      end
+
+      def deleted?(name)
+        @delete_cookies.has_key?(name)
+      end
+
+      def []=(name, value)
+        set(name, value)
+      end
+
+      def []=(name, cookie : HTTP::Cookie)
+        @cookies[name] = cookie.value
+        @set_cookies[name] = cookie
+      end
+
+      def write(headers)
+        cookies = [] of String
+        @set_cookies.each { |name, cookie| cookies << cookie.to_set_cookie_header if write_cookie?(cookie) }
+        @delete_cookies.each { |name, cookie| cookies << cookie.to_set_cookie_header }
+        headers.add("Set-Cookie", cookies)
+      end
+
+      def write_cookie?(cookie)
+        @secure || !cookie.secure
+      end
+    end
+
+    class JsonSerializer # :nodoc:
```

Comment:
With only 4096 chars using yaml instead of json saves a significant amount of space and is pretty much exactly the same code.